crates/guest-rust/rt/src/async_support.rs

```diff
@@ -397,19 +344,21 @@ unsafe fn callback_with_state(
                 subtask_drop(event1 as u32);
             }
 
-            if matches!(
-                event0,
-                EVENT_CALL_RETURNED
-                    | EVENT_STREAM_READ
-                    | EVENT_STREAM_WRITE
-                    | EVENT_FUTURE_READ
-                    | EVENT_FUTURE_WRITE
-            ) {
-                (*state).todo -= 1;
-            }
+            (*state).todo -= 1;
 
             callback_code(state, done)
         }
+
+        EVENT_STREAM_READ | EVENT_STREAM_WRITE | EVENT_FUTURE_READ | EVENT_FUTURE_WRITE => {
+            (*state).remove_waitable(event1 as u32);
+            let (waker, code) = (*state).wakers.remove(&(event1 as u32)).unwrap();
+            *code = Some(event2 as u32);
+            waker.wake();
+
+            let done = poll(state).is_ready();
+            callback_code(state, done)
+        }
```

Comment:
I'll note that in the future I want to fold `EVENT_CALL_RETURNED` into this exact same pattern as well, but that'll require some refactoring to achieve that.